### PR TITLE
Binding metadata directory in Z mode for selinux enabled docker

### PR DESCRIPTION
**Pull request by mythri-garaga**
### Summary
This change fixes the issue - https://github.com/aws/amazon-ecs-agent/issues/1113 by enabling read access to the metadata files from container processes. 

### Implementation details
- Added Info API to DockerClient interface to obtain Docker's SecurityOptions information.
- If the SecurityOptions contains selinux, we mount the metadata directory in Z mode.

### Testing

New tests cover the changes: yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/containermetadata/write_metadata_unix.go [hunk 1]
```diff
@@ -19,23 +19,38 @@ import (
...
+	tempFile              = "temp_metadata_file"
+	bindMode              = "Z"
+	selinuxSecurityOption = "selinux"
 )
 
 // createBindsEnv will do the appropriate formatting to add a new mount in a container's HostConfig
 // and add the metadata file path as an environment variable ECS_CONTAINER_METADATA_FILE
 // We add an additional uuid to the path to ensure it does not conflict with user mounts
-func createBindsEnv(binds []string, env []string, dataDirOnHost string, metadataDirectoryPath string) ([]string, []string) {
+func createBindsEnv(binds []string, env []string, dataDirOnHost string, metadataDirectoryPath string, dockerSecurityOptions []string) ([]string, []string) {
```
Comment (petderek) [inline agent/containermetadata/write_metadata_unix.go]: should we use the agent selinux config variable too?

Comment (mythri-garaga) [reply agent/containermetadata/write_metadata_unix.go]: I am not sure why we should be using agent selinux config variable.

Comment (petderek) [reply agent/containermetadata/write_metadata_unix.go]: NVM -- agent will only receive the task if the capability is set -- which implies that the config variable is set.

File: agent/engine/docker_task_engine.go [hunk 2]
```diff
@@ -1016,7 +1016,8 @@ func (engine *DockerTaskEngine) createContainer(task *apitask.Task, container *a
 	// Create metadata directory and file then populate it with common metadata of all containers of this task
 	// Afterwards add this directory to the container's mounts if file creation was successful
 	if engine.cfg.ContainerMetadataEnabled && !container.IsInternal() {
-		mderr := engine.metadataManager.Create(config, hostConfig, task, container.Name)
+		info, _ := engine.Info()
```
Comment (shubham2892) [inline agent/engine/docker_task_engine.go]: Should probably handle the error returned from `Info()` function.

File: agent/containermetadata/mocks/containermetadata_mocks.go [hunk 3]
```diff
@@ -67,17 +67,18 @@ func (mr *MockManagerMockRecorder) Clean(arg0 interface{}) *gomock.Call {
...
 	return ret0
 }
 
 // Create indicates an expected call of Create
-func (mr *MockManagerMockRecorder) Create(arg0, arg1, arg2, arg3 interface{}) *gomock.Call {
+func (mr *MockManagerMockRecorder) Create(arg0, arg1, arg2, arg3, arg4 interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Create", reflect.TypeOf((*MockManager)(nil).Create), arg0, arg1, arg2, arg3)
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Create", reflect.TypeOf((*MockManager)(nil).Create), arg0, arg1, arg2, arg3, arg4)
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Create", reflect.TypeOf((*MockManager)(nil).Create), arg0, arg1, arg2, arg3, arg4)
```
Comment (shubham2892) [inline agent/containermetadata/mocks/containermetadata_mocks.go]: Why is return called twice here?

File: agent/containermetadata/write_metadata_unix.go [hunk 4]
```diff
@@ -19,23 +19,38 @@ import (
...
+		if option == selinuxSecurityOption {
+			selinuxEnabled = true
+		}
+	}
+
 	randID := uuid.New()
 	instanceBind := fmt.Sprintf(`%s/%s:%s/%s`, dataDirOnHost, metadataDirectoryPath, mountPoint, randID)
+	if selinuxEnabled {
+		seelog.Infof("Selinux is enabled on docker, mounting data directory in Z mode")
+		instanceBind = instanceBind + ":" + bindMode
```
Comment (shubham2892) [inline agent/containermetadata/write_metadata_unix.go]: nit: Can use `fmt.sprintf()` like in next line.

File: agent/dockerclient/dockerapi/docker_client_test.go [hunk 5]
```diff
@@ -818,6 +818,58 @@ func TestDockerVersion(t *testing.T) {
...
+	info, err := client.Info(ctx, dockerclient.InfoTimeout)
+	assert.NoError(t, err)
+	assert.Equal(t, info.SecurityOptions, []string{"selinux"})
+}
+
+func TestDockerInfoError(t *testing.T) {
+	mockDockerSDK, client, _, _, _, done := dockerClientSetup(t)
+	defer done()
+
+	mockDockerSDK.EXPECT().Info(gomock.Any()).Return(types.Info{}, errors.New("Error getting info"))
```
Comment (shubham2892) [inline agent/dockerclient/dockerapi/docker_client_test.go]: Can make "Error getting info" a constant that can be used here  and in assert.

File: agent/engine/docker_task_engine_test.go [hunk 6]
```diff
@@ -725,6 +727,31 @@ func TestCreateContainerForceSave(t *testing.T) {
...
+	client.EXPECT().Info(ctx, dockerclient.InfoTimeout).Return(dockerInfo, nil)
+	metadataManager.EXPECT().Create(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), dockerInfo.SecurityOptions)
+
+	client.EXPECT().CreateContainer(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any())
+
+	metadata := taskEngine.createContainer(sleepTask, sleepContainer)
+	if metadata.Error != nil {
+		t.Error("Unexpected error", metadata.Error)
+	}
+}
```
Comment (shubham2892) [inline agent/engine/docker_task_engine_test.go]: Can use `assert.NoError`

File: agent/containermetadata/write_metadata_unix_test.go [hunk 7]
```diff
@@ -90,3 +90,35 @@ func TestWriteChmodFail(t *testing.T) {
...
+	mockMetadataDirectoryPath := ""
+	mockSecurityOptions := []string{""}
+
+	binds, _ := createBindsEnv(mockBinds, mockEnv, mockDataDirOnHost, mockMetadataDirectoryPath, mockSecurityOptions)
+
+	actualBindMode := binds[0][len(binds[0])-2:]
+	expectedBindMode := ":" + bindMode
+
+	assert.NotEqual(t, actualBindMode, expectedBindMode)
+}
```
Comment (shubham2892) [inline agent/containermetadata/write_metadata_unix_test.go]: `TestSelinuxDisabledBindMode` and `TestSelinuxEnabledBindMode` can be made [table driven](https://github.com/golang/go/wiki/TableDrivenTests). 

An example in our codebase - https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/docker_task_engine_test.go#L171

File: agent/engine/interface.go [hunk 8]
```diff
@@ -50,6 +52,8 @@ type TaskEngine interface {
 
 	Version() (string, error)
 
+	Info() (types.Info, error)
```
Comment (fenxiong) [inline agent/engine/interface.go]: i don't think this needs to be part of the TaskEngine interface, considering that this is used by the engine itself and doesn't need to be public

File: agent/containermetadata/write_metadata_unix.go (hunk 4 above)
Comment (mythri-garaga) [reply agent/containermetadata/write_metadata_unix.go]: Done.

File: agent/engine/interface.go (hunk 8 above)
Comment (mythri-garaga) [reply agent/engine/interface.go]: Agreed, removed it from askEngine interface.

File: agent/containermetadata/mocks/containermetadata_mocks.go (hunk 3 above)
Comment (mythri-garaga) [reply agent/containermetadata/mocks/containermetadata_mocks.go]: Removed.

File: agent/containermetadata/write_metadata_unix_test.go (hunk 7 above)
Comment (mythri-garaga) [reply agent/containermetadata/write_metadata_unix_test.go]: Created a table driven test for these two tests.

File: agent/engine/docker_task_engine.go (hunk 2 above)
Comment (mythri-garaga) [reply agent/engine/docker_task_engine.go]: Done. On Info() error, warning is sent and security options will be initialized to empty array.

File: agent/engine/docker_task_engine_test.go [hunk 9]
```diff
@@ -725,6 +727,48 @@ func TestCreateContainerForceSave(t *testing.T) {
...
+			error: nil,
+		},
+		{
+			name:  "Docker Info Error",
+			info:  types.Info{},
+			error: errors.New("Error getting docker info"),
+		},
+	}
+
+	client.EXPECT().APIVersion().Return(defaultDockerClientAPIVersion, nil).Times(2)
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/engine/docker_task_engine_test.go]: Resolved.

File: agent/stats/common_test.go [hunk 10]
```diff
@@ -16,6 +16,7 @@ package stats
 import (
 	"context"
 	"fmt"
+	"github.com/docker/docker/api/types"
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/stats/common_test.go]: Done.

File: agent/containermetadata/write_metadata_unix_test.go [hunk 11]
```diff
@@ -90,3 +91,40 @@ func TestWriteChmodFail(t *testing.T) {
...
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			binds, _ := createBindsEnv(mockBinds, mockEnv, mockDataDirOnHost, mockMetadataDirectoryPath, tc.securityOptions)
+			actualBindMode := binds[0][len(binds[0])-2:]
+			if tc.selinuxEnabled {
+				assert.Equal(t, actualBindMode, expectedBindMode)
+			} else {
+				assert.NotEqual(t, actualBindMode, expectedBindMode)
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/containermetadata/write_metadata_unix_test.go]: if there is no bind mode option, mount point contains generated uuid at the end of the path  so we can not set an expected value here. Hence checking for ":Z" in last two characters.

**Pull request merged**